server/control.go

```diff
@@ -97,6 +111,32 @@ func (c Controller) runLoop(ctx context.Context, leaderKey string) error {
 			return c.addon.StartWatch(ctx, addonChan)
 		})
 	}
+
+	// Watch my leadership.
+	env.Go(func(ctx context.Context) error {
+		ch := c.session.Client().Watch(ctx, leaderKey, clientv3.WithFilterPut())
+		for {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case <-c.session.Done():
+				return errors.New("session is closed")
+			case resp, ok := <-ch:
+				if !ok {
+					return errors.New("watch is closed")
+				}
+				if resp.Err() != nil {
+					return resp.Err()
+				}
+				for _, ev := range resp.Events {
+					if ev.Type == clientv3.EventTypeDelete {
+						return errors.New("leader key is deleted")
+					}
+				}
+			}
+		}
```

Comment:
This function cannot return `nil`.  In other words, this function must always fail.
Shouldn't we return `nil` in case of `ctx.Done()`? @masa213f

Reply:
@ymmt2005 
`ctx.Err()` seems to return a non-nil error, when `ctx.Done()` is closed.
So this function will not return `nil`.

https://golang.org/pkg/context/

>    // If Done is not yet closed, Err returns nil.
    // If Done is closed, Err returns a non-nil error explaining why:
    // Canceled if the context was canceled
    // or DeadlineExceeded if the context's deadline passed.
    // After Err returns a non-nil error, successive calls to Err return the same error.
    Err() error